File: cpp/problems/leetcode/0039_combination_sum/solution.cpp
// LeetCode 39
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> res;
        vector<int> path;
        function<void(int, int)> bt = [&](int start, int remain) {
            if (remain == 0) {
                res.push_back(path);
                return;
            }
            for (int i = start; i < (int)candidates.size(); ++i) {
                if (candidates[i] > remain) break;
                path.push_back(candidates[i]);
                bt(i, remain - candidates[i]);
                path.pop_back();
            }
        };
        bt(0, target);
        return res;
    }
};

int main() {
    vector<int> a{2, 3, 6, 7};
    auto out = Solution().combinationSum(a, 7);
    set<vector<int>> got(out.begin(), out.end());
    set<vector<int>> expect{{2, 2, 3}, {7}};
    assert(got == expect);
    vector<int> b{2, 3, 5};
    auto out2 = Solution().combinationSum(b, 8);
    set<vector<int>> got2(out2.begin(), out2.end());
    set<vector<int>> expect2{{2, 2, 2, 2}, {2, 3, 3}, {3, 5}};
    assert(got2 == expect2);
    cout << "0039 OK" << endl;
    return 0;
}

File: python/problems/leetcode/0039_combination_sum/solution.py
"""LeetCode 39: Combination Sum."""

from __future__ import annotations


class Solution:
    def combinationSum(
        self, candidates: list[int], target: int
    ) -> list[list[int]]:
        candidates.sort()
        res: list[list[int]] = []
        path: list[int] = []

        def bt(start: int, remain: int) -> None:
            if remain == 0:
                res.append(path.copy())
                return
            for i in range(start, len(candidates)):
                c = candidates[i]
                if c > remain:
                    break
                path.append(c)
                bt(i, remain - c)
                path.pop()

        bt(0, target)
        return res


if __name__ == "__main__":
    out = Solution().combinationSum([2, 3, 6, 7], 7)
    assert sorted(map(tuple, out)) == [(2, 2, 3), (7,)]
    out2 = Solution().combinationSum([2, 3, 5], 8)
    assert sorted(map(tuple, out2)) == [(2, 2, 2, 2), (2, 3, 3), (3, 5)]
    print("0039 OK")
